components/setup/DpuSetup.tsx

```diff
@@ -0,0 +1,196 @@
+import React, { useState, useEffect } from 'react';
+import axios from 'axios';
+import type { Session } from "next-auth";
+
+import { CopyToClipboard } from 'react-copy-to-clipboard';
+import { MdContentCopy } from "react-icons/md";
+import { getUserRepositories } from '../../utils/providerAPI/getUserRepositories';
+import Button from '../Button';
+
+interface DpuSetupProps {
+	userId: string;
+	selectedProvider: string;
+	selectedInstallationType: string;
+	session: Session | null;
+}
+
+interface SaveSetupReposApiBodyArgs {
+	owner: string,
+	provider: string,
+	repos: string [],
+	installationId: string
+}
+
+export type RepoIdentifier = { repo_provider: string, repo_owner: string, repo_name: string };
+
+async function filterProviderRepos(session: Session, targetProvider: string) {
+	const allRepos = await getUserRepositories(session);
+	// Filter repos based on the targetProvider
+	const filteredRepos = allRepos.filter(repo => repo.repo_provider === targetProvider);
+	return filteredRepos;
+}
+
+function formatRepoListInSaveSetupArgsForm(repos: RepoIdentifier[], install_id: string) {
+	// Group by repo_owner and generate setup args
+	const setupArgsMap: Map<string, SaveSetupReposApiBodyArgs> = new Map();
+	repos.forEach(repo => {
+		const key = repo.repo_owner;
+		if (!setupArgsMap.has(key)) {
+			setupArgsMap.set(key, {
+				owner: repo.repo_owner,
+				provider: repo.repo_provider,
+				repos: [],
+				installationId: install_id
+			});
+		}
+		setupArgsMap.get(key)!.repos.push(repo.repo_name);
```

Comment:
Can you explain the use of the `!.` operation instead of the safer `?.` operation?